### DISCUSSION TITLE
JSON output for real time strip chart plotting

### DISCUSSION THREAD

**Discussion by UcfAerokid** [JSBSim-Team/jsbsim#580] [Show and tell]
[jsbsimJsonOut.zip](https://github.com/JSBSim-Team/jsbsim/files/7991802/jsbsimJsonOut.zip)
I was able to get JSON format output working to feed the PlotJuggler live plotting utility.  It outputs JSON strings to a socket where PlotJuggler can pick up the object format and parse the object tree to run realtime strip charts (see screen cap attached). 

![jsbsimLivePlot](https://user-images.githubusercontent.com/95836100/152281999-ab877b73-75f9-4b79-adbe-f744b4edcd44.png)

 I used the FgOutputSocket.cpp/h as a starting point and it reads the output element as seen below:

```xml
 <output name="localhost" type="JSON" port="49870" rate="10">
    <simulation> ON </simulation>
    <atmosphere> ON </atmosphere>
    <massprops> OFF</massprops>
```

 etc...

I don't have the individual properties working just yet, but that will be the next step when I get another chance to work on it again.  They will be appended at the end after all the sections that are turned ON have been included.

The .zip also has the Wireshark capture showing the strings.  Select decode as JSON and WS will display the JSON object tree.

Any feedback or ideas greatly appreciated.  I don't have it all working and I'm a Git newb, so I'm not sure how to submit the code, but it's in the .zip file for your consideration.

Thanks @seanmcleod for pointing me in the direction of PlotJuggler.

**Comment by seanmcleod** [comment]
@UcfAerokid thanks, the json output for PlotJuggler has been on my todo list for ages 😉 

But yes you'll need to get up to speed with git though to enable other JSBSim developers to be able to easily review your code via a github pull request.

Here is a very simple step by step guide on how to go about submitting a pull request, starting with forking the JSBSim repo, creating a branch etc.

https://jarv.is/notes/how-to-pull-request-fork-github/

P.S
Formatting tip for comments, take a look at how I edited your original comment above to use the triple back-tick with xml (c++ and others also supported) specified in order to get better formatting, i.e. 

![image](https://user-images.githubusercontent.com/824856/152288570-2104f340-66bf-4363-ba30-d7010161333e.png)

**Comment by seanmcleod** [comment]
Just looking at your sample `<output type="JSON".... >` I'm wondering if we need to think about not overloading the `type` element and split out the output format, i.e. `json`, `csv` etc. from the output stream type, i.e. `local file`, `socket` etc.

So at the moment the `type` element effectively specifies both the format of the data that is output and also the output stream type. For example what if I want to output `JSON` but to a file and not to a socket?

Current situation.

```xml
<output name="JSBout172C.csv" type="CSV" rate="60">

<output name="localhost" type="SOCKET" port="5138" protocol="UDP" rate="10">

<output name="localhost" type="JSON" port="49870" rate="10">

```

**Comment by seanmcleod** [comment]
So potentially something along these lines?

The `type` element specifies the output stream type, i.e. `FILE | SOCKET` and a new element `format` specifies the format of the data, i.e. `CSV | JSON`.

We'd need some backwards compatibility, so if `type = CSV` then assume `type = FILE` and `format = CSV` and if `type = SOCKET` and no `format` element then assume `format = CSV`. 

Actually @dpculp's QTJSBSim introduced another `type = QTJSBSIM` which is a combination of a socket and particular data format.

**Comment by bcoconni** [comment]
> I was able to get JSON format output working to feed the PlotJuggler live plotting utility. It outputs JSON strings to a socket where PlotJuggler can pick up the object format and parse the object tree to run realtime strip charts (see screen cap attached).

@UcfAerokid Nice ! A pull request for this feature will be most appreciated as @seanmcleod said.

**Comment by UcfAerokid** [comment]
OK @bcoconni @seanmcleod .  If I got this Git thing right there should be an active branch available to pull from.  Please let me know if you see it and it's useful, or any feedback on the code, etc.

I've used Visual SourceSafe, SubVersion, Mercurial, Razor etc. throughout my career, but it's my first time dipping my toes into Git, so please let me know if I need to do something different.  I followed your link with the suggested procedures as best as I could from my GitHub app.

Thanks.
UcfAerokid
